File: swiftbible/Models/HistoryModels.swift
//
//  HistoryModels.swift
//  swiftbible
//
//  Models for the History tab — sections, articles, body blocks,
//  pull quotes, and external sources.
//

import Foundation

struct HistorySection: Identifiable, Hashable {
    let id: String
    let title: String
    let subtitle: String
    let symbol: String
    let era: String
    let articles: [HistoryArticle]
}

struct HistoryArticle: Identifiable, Hashable {
    let id: String
    let title: String
    let subtitle: String
    let era: String
    let estimatedMinutes: Int
    let body: [BodyBlock]
    let pullQuotes: [PullQuote]
    let sources: [HistorySource]
    let related: [String]
}

enum BodyBlock: Hashable {
    case paragraph(String)
    case heading(String)
    case quote(text: String, attribution: String)
    case list([String])
    case timeline([TimelineEntry])
    case divider
}

struct TimelineEntry: Hashable, Identifiable {
    let id = UUID()
    let year: String
    let event: String
}

struct PullQuote: Hashable, Identifiable {
    let id = UUID()
    let text: String
    let attribution: String
    let context: String?
}

struct HistorySource: Hashable, Identifiable {
    let id = UUID()
    let title: String
    let author: String?
    let kind: SourceKind
    let url: String?
    let note: String?
}

enum SourceKind: String, Hashable {
    case primary
    case scholarly
    case encyclopedia
    case scripture

    var label: String {
        switch self {
        case .primary: return "Primary"
        case .scholarly: return "Scholarly"
        case .encyclopedia: return "Reference"
        case .scripture: return "Scripture"
        }
    }
}

File: android/app/src/main/java/biz/am2/swiftbible/data/history/HistoryModels.kt
package biz.am2.swiftbible.data.history

data class HistorySection(
    val id: String,
    val title: String,
    val subtitle: String,
    val era: String,
    val emoji: String,
    val articles: List<HistoryArticle>,
)

data class HistoryArticle(
    val id: String,
    val title: String,
    val subtitle: String,
    val era: String,
    val estimatedMinutes: Int,
    val body: List<BodyBlock>,
    val pullQuotes: List<PullQuote> = emptyList(),
    val sources: List<HistorySource> = emptyList(),
    val related: List<String> = emptyList(),
)

sealed class BodyBlock {
    data class Paragraph(val text: String) : BodyBlock()
    data class Heading(val text: String) : BodyBlock()
    data class Quote(val text: String, val attribution: String) : BodyBlock()
    data class BulletList(val items: List<String>) : BodyBlock()
    data class Timeline(val entries: List<TimelineEntry>) : BodyBlock()
    object Divider : BodyBlock()
}

data class TimelineEntry(val year: String, val event: String)

data class PullQuote(
    val text: String,
    val attribution: String,
    val context: String? = null,
)

data class HistorySource(
    val title: String,
    val author: String? = null,
    val kind: SourceKind,
    val url: String? = null,
    val note: String? = null,
)

enum class SourceKind(val label: String) {
    PRIMARY("Primary"),
    SCHOLARLY("Scholarly"),
    ENCYCLOPEDIA("Reference"),
    SCRIPTURE("Scripture"),
}
